[Codegen][LegalizeIntegerTypes] Improve shift through stack

futog opened the pull request:
Minor improvement on cc39c3b17fb2598e20ca0854f9fe6d69169d85c7.

Use an aligned stack slot to store the shifted value.
Use the native register width as shifting unit, so the load of the
shift result is aligned.

If the shift amount is a multiple of the native register width, there is
no need to do a follow-up shift after the load. I added new tests for
these cases.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -4530,14 +4530,25 @@ void DAGTypeLegalizer::ExpandIntRes_ShiftThroughStack(SDNode *N, SDValue &Lo,
-  bool ShiftByByteMultiple =
-      DAG.computeKnownBits(ShAmt).countMinTrailingZeros() >= 3;
+  EVT LoadStoreVT = VT;
+  do {
+      LoadStoreVT = TLI.getTypeToTransformTo(*DAG.getContext(), LoadStoreVT);
+  }while (!TLI.isTypeLegal(LoadStoreVT));
```
Formatting and spacing

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -4577,16 +4587,22 @@ void DAGTypeLegalizer::ExpandIntRes_ShiftThroughStack(SDNode *N, SDValue &Lo,
-  SDValue ByteOffset = DAG.getNode(ISD::SRL, dl, ShAmtVT, ShAmt,
-                                   DAG.getConstant(3, dl, ShAmtVT), Flags);
+  SDValue OffsetInBits = DAG.getNode(ISD::SHL, dl, ShAmtVT,
+                                     DAG.getNode(ISD::SRL, dl, ShAmtVT, ShAmt, DAG.getConstant(Log2_32(ShiftUnitInBits), dl, ShAmtVT), Flags),
+                                     DAG.getConstant(Log2_32(ShiftUnitInBits), dl, ShAmtVT));
```
Use some temporary variables and reduce indentation

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -4530,14 +4530,25 @@ void DAGTypeLegalizer::ExpandIntRes_ShiftThroughStack(SDNode *N, SDValue &Lo,
+  }while (!TLI.isTypeLegal(LoadStoreVT));
+
+  const Align LoadStoreAlign = [&]() -> Align {
+      if (TLI.allowsMisalignedMemoryAccesses(LoadStoreVT))
+          return Align(1);
```
Don't understand this. Just because it's allowed doesn't mean it's a good idea to use Align 1

futog replied:
I did not wanted to change the behavior for the targets which supports unaligned memory access. In retrospect, this was not a good idea. Also I found this comment on the original commit by efriedma (https://reviews.llvm.org/D140638):

> It might be worth implementing a strategy that avoids unaligned loads (by splitting the shift amount by the native register width instead of CHAR_BIT). On targets that don't have native unaligned loads, they're pretty expensive. Even on targets that do have unaligned loads, an aligned load can reduce the cost of the store forwarding stall. (But on targets with fast unaligned loads, they're probably worth using if the shift amount is known to be a multiple of CHAR_BIT.)

So I will implement this instead.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -4530,14 +4530,35 @@ void DAGTypeLegalizer::ExpandIntRes_ShiftThroughStack(SDNode *N, SDValue &Lo,
+  } while (!TLI.isTypeLegal(LoadStoreVT));
+
+  const unsigned KnownTrailingZeros =
+      DAG.computeKnownBits(ShAmt).countMinTrailingZeros();
```
Defer the computeKnownBits call until absolutely necessary

futog replied:
I cannot defer it any more, at least not nicely. I use it to get the correct alignment.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -4530,14 +4530,35 @@ void DAGTypeLegalizer::ExpandIntRes_ShiftThroughStack(SDNode *N, SDValue &Lo,
-      DAG.computeKnownBits(ShAmt).countMinTrailingZeros() >= 3;
+  EVT LoadStoreVT = VT;
+  do {
+    LoadStoreVT = TLI.getTypeToTransformTo(*DAG.getContext(), LoadStoreVT);
```
I think getTypeToTransform will return a legal type on the first call. Do you hit any cases where this actually loops?

futog replied:
`ExpandIntRes_ShiftThroughStack` is called only if the result must be expanded in at least two steps (See https://github.com/llvm/llvm-project/blob/0f111ba790a28056d42a0ae376371d7b694454ae/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp#L4668).
Also TargetLowering::isTypeLegal returns true if the target has native support for the specified value type.
On riscv32 for i128 shift it will loop.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -4530,14 +4530,35 @@ void DAGTypeLegalizer::ExpandIntRes_ShiftThroughStack(SDNode *N, SDValue &Lo,
+    unsigned IsFast = 0;
+    const bool AllowsFastMisalignedMemoryAccesses =
+        TLI.allowsMisalignedMemoryAccesses(
+            LoadStoreVT, /*AddrSpace*/ 0, /*Alignment*/ Align(1),
```
Do not assume address space 0. Use the alloca address space. 

Also parameter name comments should end in a =. Also why Align(1)?

futog replied:
also added `MachineMemOperand::MOLoad | MachineMemOperand::MOStore` for Flags

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -4530,14 +4530,35 @@ void DAGTypeLegalizer::ExpandIntRes_ShiftThroughStack(SDNode *N, SDValue &Lo,
+    const bool AllowsFastMisalignedMemoryAccesses =
+        TLI.allowsMisalignedMemoryAccesses(
+            LoadStoreVT, /*AddrSpace=*/DAG.getDataLayout().getAllocaAddrSpace(),
+            /*Alignment=*/Align(LoadStoreVT.getStoreSize()),
```
If you're going to use the natural alignment, it doesn't make sense to call TLI.allowsMisalignedMemoryAccesses in the first place. You would need to pull this from the value you are going to create the actual load/store with

futog replied:
Sorry but I don't get it... This function is for checking if I should use natural alignment or not. If the target supports fast unaligned memory access and the shift is a multiple of CHAR_BIT, I don't use natural alignment.

You mean that I should drop this approach and always use `DAG.getReducedAlign(LoadStoreVT, /*UseABI=*/false);`?

arsenm replied:
I mean I don't understand why you are considering unaligned access. This code is creating a temporary stack slot, that can be whatever alignment you choose. Deciding to under-align this stack slot is kind of going out of scope for what this code is doing. Can you just get rid of all this legal alignment checking business and just use the naturally aligned type slot? 

If we wanted an optimization to under-align stack slots, it ideally would be a separate optimization checking the uses of the stack slot instead of just performing it in this one specific legalization step.

That is, delete all of this LoadStoreAlign code, call CreateStackTemproary as below and use the alignment of whatever it chose

futog replied:
thanks for the time you are investing in this, I uploaded a new version

arsenm commented on llvm/test/CodeGen/AArch64/wide-scalar-shift-by-byte-multiple-legalization.ll:
```diff
@@ -179,21 +179,36 @@ define void @ashr_16bytes(ptr %src.ptr, ptr %byteOff.ptr, ptr %dst) nounwind {
-; ALL-NEXT:    sub sp, sp, #64
-; ALL-NEXT:    ldp x9, x8, [x0, #16]
-; ALL-NEXT:    movi v0.2d, #0000000000000000
+; ALL-NEXT:    ldr q0, [x0]
```
The AArch64 result looks much worse

futog replied:
the reason of the regression is that in the legacy the shift unit was 1 byte, and in this test the shift is a multiple of 1 byte, so we could figure out that the shift will be one load only.

Now the shift unit is 8 byte, so the shift will be load + shift

arsenm replied:
You should take the alignment from the stack temporary, which is not 1

futog replied:
did you mean to leave this comment here?...

arsenm commented on llvm/test/CodeGen/RISCV/wide-scalar-shift-by-byte-multiple-legalization.ll:
```diff
@@ -723,98 +723,117 @@ define void @lshr_16bytes(ptr %src.ptr, ptr %byteOff.ptr, ptr %dst) nounwind {
-; RV32I-NEXT:    lbu a3, 0(a0)
-; RV32I-NEXT:    lbu a4, 1(a0)
+; RV32I-NEXT:    addi sp, sp, -32
+; RV32I-NEXT:    lbu a3, 1(a0)
```
This looks worse?

futog replied:
the same as for aarch64, plus it seems that the load/store alignment used in the test, which is 1, does not mix well with the legalized code.

arsenm replied:
I think it's strange this test is using an unaligned load in the first place, but this shouldn't be regressing the load that's unrelated to the use shift

futog replied:
I checked it again, and the reason is the same as in for aarch64. We need a load+shift here.

futog replied:
I will add some tests where the shift amount is known to be a multiple of 4 bytes, so 1 load with an offset is enough to do the shift.

futog replied:
The test uses unaligned load/store for the shiftee (which lowers to LBUs), but we use aligned load/store for stack operations, which lowers to LWs. This means that the registers used for loading the shiftee cannot be used directly for storing the shiftee on the stack, and the registers used for loading the shifted value from the stack cannot be used directly to store the result.

LBU: loads a 8-bit value from memory, then sign-extends to 32-bits before storing in the destination register
LW: loads a 32-bit value from memory into the destination register

arsenm commented on llvm/test/CodeGen/X86/wide-scalar-shift-legalization.ll:
```diff
@@ -5366,454 +4833,370 @@ define void @shl_64bytes(ptr %src.ptr, ptr %bitOff.ptr, ptr %dst) nounwind {
+; X86-HAVE-BMI2-NO-SHLD-NEXT:    movl (%ebp), %ebp
+; X86-HAVE-BMI2-NO-SHLD-NEXT:    xorps %xmm0, %xmm0
+; X86-HAVE-BMI2-NO-SHLD-NEXT:    movaps %xmm0, {{[0-9]+}}(%esp)
+; X86-HAVE-BMI2-NO-SHLD-NEXT:    movl %eax, {{[0-9]+}}(%esp)
```
x86 looks plausibly better?

arsenm commented on llvm/test/CodeGen/PowerPC/ctrloop-sh.ll:
```diff
@@ -159,59 +147,53 @@ define void @foo3(ptr %a, ptr readonly %b, ptr readonly %c) #0 {
+; CHECK-NEXT:    stw 11, 44(1)
+; CHECK-NEXT:    clrlwi 11, 12, 27
+; CHECK-NEXT:    stw 10, 36(1)
+; CHECK-NEXT:    subfic 0, 11, 32
```
I think this is better

arsenm commented on llvm/test/CodeGen/Mips/llvm-ir/lshr.ll:
```diff
@@ -398,197 +398,159 @@ define signext i128 @lshr_i128(i128 signext %a, i128 signext %b) {
-; MIPS2-NEXT:    lwl $7, 12($1)
-; MIPS2-NEXT:    lwr $7, 15($1)
-; MIPS2-NEXT:    srlv $1, $7, $2
+; MIPS2-NEXT:    lw $1, 60($sp)
```
This is probably better

arsenm reviewed (commented):
I think the code change looks fine, but the AArch64 and RISCV cases look like regressions. Is there an explanation for those? Is the type transform loop excessively widening these now?

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -4603,23 +4616,23 @@ void DAGTypeLegalizer::ExpandIntRes_ShiftThroughStack(SDNode *N, SDValue &Lo,
       VT, dl, Ch, AdjStackPtr,
-      MachinePointerInfo::getUnknownStack(DAG.getMachineFunction()), Align(1));
-  // We've performed the shift by a CHAR_BIT * [_ShAmt / CHAR_BIT_]
+      MachinePointerInfo::getUnknownStack(DAG.getMachineFunction()));
```
I think we have a way to express fixed stack with unknown offset, but you should be setting the alignment here from base + offset

futog replied:
I guess I can use `getReducedAlign(LoadStoreVT, false)` here, given that the shift unit is based on this type. Is this ok?

arsenm replied:
I don't think that's correct. It should be the common alignment of the base alignment and the element size

futog commented:
Hi @arsenm, do you have any more concerns with this PR?

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -4551,10 +4560,11 @@ void DAGTypeLegalizer::ExpandIntRes_ShiftThroughStack(SDNode *N, SDValue &Lo,
-  SDValue StackPtr = DAG.CreateStackTemporary(
-      TypeSize::getFixed(StackSlotByteWidth), StackSlotAlignment);
+  Align StackAlign = DAG.getReducedAlign(StackSlotVT, /*UseABI=*/false);
+  assert(DAG.getReducedAlign(LoadVT, /*UseABI=*/false) <= StackAlign);
```
Don't see the point of this assert

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -4603,23 +4622,24 @@ void DAGTypeLegalizer::ExpandIntRes_ShiftThroughStack(SDNode *N, SDValue &Lo,
+  SDValue Res =
+      DAG.getLoad(VT, dl, Ch, AdjStackPtr,
+                  MachinePointerInfo::getUnknownStack(DAG.getMachineFunction()),
+                  DAG.getReducedAlign(LoadVT, /*UseABI=*/false));
```
Reuse the alignment from above?

futog replied:
changed it to use commonAlignment

futog replied:
sorry should not have put it into resolved

arsenm commented on llvm/test/CodeGen/AArch64/wide-scalar-shift-by-byte-multiple-legalization.ll:
```diff
@@ -238,18 +270,34 @@ define void @ashr_32bytes(ptr %src.ptr, ptr %byteOff.ptr, ptr %dst) nounwind {
-; ALL-NEXT:    ldr q0, [x8]
-; ALL-NEXT:    str q0, [x2]
-; ALL-NEXT:    stp x10, x9, [x2, #16]
+; ALL-NEXT:    lsl x8, x10, #3
```
This is a lot worse?

futog replied:
In the legacy code the shift unit is 1 byte. If the shift amount is known to be a multiple of 1 byte, one load with an offset is enough to do the shift. In this test case the shift amount is the multiple of 1 byte.

Now the shift unit for aarch64 is 8 byte, and the shift amount is not a multiple of 8 bytes, so we need a load+shift operation.

I think that this could be fixed only if we let unaligned loads in some cases.

davemgreen replied:
If this is a lot worse then why not just allow unaligned loads? It could test allowsMisalignedMemoryAccesses to check if they are supported and the IsFast parameter can tell you if they are efficient.

arsenm replied:
The unaligned load is on the unrelated operation that happened to produce the incoming value. There's no natural reason in context to care about unaligned loads

davemgreen replied:
I see, it was in a previous version but taken out. So this makes strict-align better, but systems that allow unaligned access can be worse if the shift amount is known to be at least byte aligned.

futog replied:
Maybe we could add a logic the calculate the shifting unit based on the support of unaligned memory access. I was thinking something like this:
`
  Align LoadAlign = [&]() {
    EVT LoadVT = VT;
    do {
      LoadVT = TLI.getTypeToTransformTo(*DAG.getContext(), LoadVT);
    } while (!TLI.isTypeLegal(LoadVT));

    Align LoadAlign(LoadVT.getStoreSize());
    for (Align Tmp = LoadAlign;; Tmp = Tmp.previous()) {
      unsigned IsFast = 0;
      const bool AllowsFastMisalignedMemoryAccesses =
          TLI.allowsMisalignedMemoryAccesses(
              LoadVT, DAG.getDataLayout().getAllocaAddrSpace(), Tmp,
              MachineMemOperand::MOLoad, &IsFast) &&
          IsFast;
      if (not AllowsFastMisalignedMemoryAccesses) {
        break;
      }
      LoadAlign = Tmp;
      if (LoadAlign == Align(1))
        break;
    }
    return LoadAlign;
  }();
  const unsigned ShiftUnitInBits = LoadAlign.value() * 8;
  ...
  // And load it! While the load is not legal, legalizing it is obvious.
  SDValue Res =
      DAG.getLoad(VT, dl, Ch, AdjStackPtr,
                  MachinePointerInfo::getUnknownStack(DAG.getMachineFunction()),
                  commonAlignment(StackAlign, LoadAlign.value()));


`
@arsenm I know that you have already approved the PR, but what do you think?

arsenm replied:
It still doesn't make sense to be to be considering unaligned accesses when there is no unaligned access in the source operation, and no need to introduce one. I don't think anything like this belongs in this patch, and it seems more like a missing optimization elsewhere

davemgreen replied:
I'm not sure what is meant by "no need to introduce one". Is it not just a better optimization to to be able to load unaligned back from the stack?

arsenm replied:
The original operation has no stack operation. This is creating a new stack slot that can use any alignment, so why would you introduce a new underaligned slot? I still think any adjustment here would be best done in a follow up PR

davemgreen replied:
Am I right is saying that we start off with a large i256 shift, where we can work out that it shifts by a multiple of 8? So we store the original value to the stack slot of twice the size, and reload from `stack+shift/8`? And if you use that alignment you don't need to do anything else, but if you require a higher alignment then you might need to do a follow-up shift too?

futog replied:
I'm computing `shift_aligned` value from the load alignment. Then I load from `stack+shift/shift_aligned`, which will be an aligned load now. If I can work out that `shift` is a multiple of `shift_aligned`, we are done, otherwise we need a follow-up shift.

davemgreen replied:
Yeah OK. I don't see why we would make this optimization worse by requiring aligned loads on machines where there is no penalty for doing so, but it is relatively unlikely to come up in practice so feel free to ignore me and continue on getting this committed.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -4574,20 +4584,29 @@ void DAGTypeLegalizer::ExpandIntRes_ShiftThroughStack(SDNode *N, SDValue &Lo,
+
+  Flags.setExact(true);
+  SDValue Offset = DAG.getNode(ISD::SRL, dl, ShAmtVT, OffsetInBits,
+                               DAG.getConstant(3, dl, ShAmtVT), Flags);
```
This is a sneaky hardcoded byte size. Is there a way to make it more obvious this is a byte shift?

bjope commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -4530,14 +4530,23 @@ void DAGTypeLegalizer::ExpandIntRes_ShiftThroughStack(SDNode *N, SDValue &Lo,
+    LoadVT = TLI.getTypeToTransformTo(*DAG.getContext(), LoadVT);
+  } while (!TLI.isTypeLegal(LoadVT));
+
+  const unsigned ShiftUnitInBits = LoadVT.getStoreSize() * 8;
```
I think this could use getStoreSizeInBits().

(Unless you want to highlight that it needs to match with all the 8's and 3's below for some reason.)

futog replied:
I will also replace the 8s whith CHAR_BIT

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -4574,20 +4584,29 @@ void DAGTypeLegalizer::ExpandIntRes_ShiftThroughStack(SDNode *N, SDValue &Lo,
   }
   // And spill it into the stack slot.
-  Ch = DAG.getStore(Ch, dl, Init, StackPtr, StackPtrInfo, StackSlotAlignment);
+  Ch = DAG.getStore(Ch, dl, Init, StackPtr, StackPtrInfo);
```
This store is not getting the correct alignment set. The stack alignment should be the next argument if using the MachinePointerInfo overloads

futog replied:
Alignment is calculated by getEVTAlign inside the function, which gives the correct alignment in this case or? Or do you mean that I should I add StackAlign explicitly?

arsenm replied:
Magically figuring it out inside this is more confusing than just explicitly passing it, especially since you already have alignment logic here

futog committed:
[Codegen][LegalizeIntegerTypes] Improve shift through stack

Minor improvement on cc39c3b17fb2598e20ca0854f9fe6d69169d85c7.

Use an aligned stack slot to store the shifted value.
Use the native register width as shifting unit, so the load of the
shift result is aligned.

If the shift amount is a multiple of the native register width, there is
no need to do a follow-up shift after the load. I added new tests for
these cases.

arsenm commented on llvm/test/CodeGen/AArch64/wide-scalar-shift-by-byte-multiple-legalization.ll:
```diff
@@ -186,23 +186,68 @@ define void @lshr_32bytes(ptr %src.ptr, ptr %byteOff.ptr, ptr %dst) nounwind {
+; ALL-NEXT:    stp x8, x9, [x2]
+; ALL-NEXT:    add sp, sp, #64
+; ALL-NEXT:    ret
+  %src = load i256, ptr %src.ptr, align 1
```
Precommit the new tests? I can't easily see what the diff is anymore

futog replied:
You mean add it as a first commit to this PR so you can filter it out? Or a separate PR.

arsenm replied:
Either way, or just directly push them

futog commented:
Hi @arsenm,

Do you have any suggestion on how to proceed with this PR? I think every obvious additional instructions are explained now.

Regards,
Gergely

arsenm reviewed (approved):
Still don't really get why the tests were using unaligned loads in the first place, or why this regresses them, but it shouldn't be directly related to this patch

topperc commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -4666,15 +4681,16 @@ void DAGTypeLegalizer::ExpandIntRes_ShiftThroughStack(SDNode *N, SDValue &Lo,
+  SDValue Res =
+      DAG.getLoad(VT, dl, Ch, AdjStackPtr,
+                  MachinePointerInfo::getUnknownStack(DAG.getMachineFunction()),
+                  commonAlignment(StackAlign, LoadVT.getStoreSize()));
```
Doesn't this alignment need to take into account the byte offset?

topperc replied:
Nevermind that's factored into the LoadVT.

RKSimon commented:
@futog please can you rebase?

futog commented:
@arsenm could you please merge? I don't have the right to do so. I only squashed the commits into one.
